File: 1658-minimum-operations-to-reduce-x-to-zero/1658-minimum-operations-to-reduce-x-to-zero.py
class Solution:
    def minOperations(self, nums: list[int], x: int) -> int:
        total = 0
        for num in nums:
            total += num
        
        target = total - x
        if target == 0:
            return len(nums)

        longest = 0
        window_sum = 0
        left = 0
        for right in range(len(nums)):
            window_sum += nums[right]
            while left < len(nums) and window_sum > target:
                window_sum -= nums[left]
                left += 1
            if window_sum == target:
                longest = max(longest, right - left + 1)
        
        return -1 if longest == 0 else len(nums) - longest

# array, sliding window
# time: O(n)
# space: O(1)

File: 1658-minimum-operations-to-reduce-x-to-zero/1658-minimum-operations-to-reduce-x-to-zero.go
func minOperations(nums []int, x int) int {
    total := 0
    for _, num := range nums {
        total += num
    }

    target := total - x
    longest := 0
    windowSum := 0
    left := 0

    if target == 0 {
        return len(nums)
    }

    for right := 0; right < len(nums); right++ {
        windowSum += nums[right]

        for windowSum > target && left < len(nums) {
            windowSum -= nums[left]
            left++
        }

        if windowSum == target {
            longest = max(longest, right-left+1)
        }
    }

    if longest == 0 {
        return -1
    }

    return len(nums) - longest
}

// array, sliding window
// time: O(n)
// space: O(1)